Ensure `reload` sets correct owner for each association (ActiveRecord::Persistence)

dmytro-savochkin opened the pull request:
### Motivation / Background

Fixes #46363 which was caused by https://github.com/rails/rails/pull/41112

Currently calling `reload` re-calculates the association cache based on the object freshly loaded from the database. The fresh object gets assigned as `owner` for each association in the association cache. That object however is not the same as `self` (even though they both point to the same record). Under some specific circumstances (using `has_one/has_many through` associations and transactions with `after_commit` callbacks) this might lead to losing the effects of assignments done inside the callback.

This commit fixes it by making `reload` point to `self` as `owner` for each association in the association cache.

I think this is safe because as far as I can see in https://github.com/rails/rails/blob/main/activerecord/lib/active_record/associations.rb#L320 it always passes `self` when creating a new association.

### Detail

This Pull Request changes `ActiveRecord::Persistence#reload` method.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [x] CI is passing.

dmytro-savochkin commented:
@eileencodes, may I ask you to review please (since you were adding association cache line before)? Thanks!

dmytro-savochkin commented:
Hey @eileencodes! Sorry to ping again but is it possible for you to review this? :)

Or, alternatively, I would be quite happy if anyone else reviewed this.

nvasilevski commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -971,6 +971,7 @@ def reload(options = nil)
       end
 
       @association_cache = fresh_object.instance_variable_get(:@association_cache)
+      @association_cache.each_value { |association| association.instance_variable_set(:@owner, self) }
```
I wonder what core team folks think on that but I always feel uncomfortable about using `instance_variable_set` unless we are patching some code we have no access to. Otherwise we are using something that wasn't supposed to be used even within the framework itself and makes me think whether we should add a public method to `association` to be able to set the `owner`. To be fair it might not be a solid argument as we already set and access instance variables on the lines above and below this one 🙃 

Other than that, the PR makes sense to me.

dmytro-savochkin replied:
Yes, this was exactly my thinking!

I am also not very comfortable using `instance_variable_get/instance_variable_set` but in this particular case I decided to go with it because:
1. This is already done one line above and below just as you said :)
2. I wasn't completely sure my approach would be accepted (dunno, maybe some other reasons I am not aware of) so doing some major change, like trying to introduce some new public method into `association` (which in turn might require changes to the docs, etc, etc) so that it can be used in this place, seemed a bit too premature to me.

dmytro-savochkin replied:
But if you think it's worth it, let me know, and I will try to introduce my change via a new public method. 
However the lines above and below will still be using `instance_variable_get`, so dunno... Changing those in this PR as well seems like a bit too much for such a small problem/PR.

rafaelfranca replied:
Yeah, let's avoid those instance variable set here. If we need to set the owner outside of that object, we should expose a writter.

dmytro-savochkin replied:
No problem, updated.

dmytro-savochkin commented:
Hey @nvasilevski!
Sorry for bothering but maybe you know who could be added to reviewers to this?
I understand sometimes it might take a few months here for a PR to be addressed but until I am completely sure it's the only option I'd like to try fasten things up a little :)

skipkayhil commented:
Please don't ping individuals for review unless they have asked you to. The best way to get feedback would be asking in the [Discord](https://rubyonrails.org/2022/6/13/rails-discord-server-is-now-open-to-the-public)

dmytro-savochkin commented:
> Please don't ping individuals for review unless they have asked you to. The best way to get feedback would be asking in the [Discord](https://rubyonrails.org/2022/6/13/rails-discord-server-is-now-open-to-the-public)

Oh, wow, there is also a Discord! Thanks, I will try that.
So far however I've followed [this guide for feedback](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#get-some-feedback) and on the only thing mentioned there (ror-core mailing list) there was no reply.

dmytro-savochkin commented:
This PR is still relevant for Rails 7.1.1, the issue is still there.

rafaelfranca commented on activerecord/test/cases/persistence/reload_association_cache_test.rb:
```diff
@@ -0,0 +1,47 @@
+
+require "cases/helper"
+
+# The test case below is aimed to fix a bug that was noticeable only with a very specific
+# setup of models. The models are fully defined inside the test case so that they are not
+# accidentally changed later causing the bug effects to disappear.
```
Let's move this test to where it makes sense. You can define new models by creating new classes in the test/models folder, no need to do what is being done here.

Those models here still leaks to all the other files, so you aren't really isolating those models from other tests

dmytro-savochkin replied:
Yep, updated as well.

dmytro-savochkin committed:
Ensure `reload` sets correct owner for each association

Fix #46363.

Currently calling `reload` re-calculates the association cache based
on the object freshly loaded from the database. The fresh object gets
assigned as `owner` for each association in the association cache.
That object however is not the same as `self` (even though they
both point to the same record). Under some specific circumstances
(using `has_one/has_many through` associations and transactions with
`after_commit` callbacks) this might lead to losing the effects of
assignments done inside the callback.

This commit fixes it by making `reload` point to `self` as `owner`
for each association in the association cache.
